Add 16A8W linear ops support and test

Ninja91 opened the pull request:
Summary:
- Adds linear ops test using the 16A8W config in INT16 profile.
- Adds support in view ops validation for INT16 Dtype.
- Validated with TOSA pipeline test.

Note: Not verified with tosa reference model run.

Differential Revision: D80308822




cc @digantdesai @freddan80 @per @zingo @oscarandersson8218

digantdesai commented on backends/arm/operators/op_view.py:
```diff
@@ -44,7 +44,7 @@ def define_node(
             self.target,
             [inputs[0], output],
-            [ts.DType.INT8, ts.DType.INT32, ts.DType.FP32, ts.DType.BOOL],
+            [ts.DType.INT8, ts.DType.INT16, ts.DType.INT32, ts.DType.FP32, ts.DType.BOOL],
```
Should we check tosa_spec before appending INT16 here?

digantdesai commented on backends/arm/quantizer/arm_quantizer.py:
```diff
@@ -143,6 +143,111 @@ def get_symmetric_quantization_config(
 
 
+@functools.lru_cache
+def get_16a8w_quantization_config(
```
Instead of duplicating, should we get `quantization_config` `get_symmetric_quantization_config` and update just the activation spec in this function?

Ninja91 replied:
Addressed.

digantdesai commented on backends/arm/test/ops/test_linear.py:
```diff
@@ -258,3 +261,34 @@ def test_linear_vgf_INT(test_data: torch.Tensor):
+        quantization_config=get_16a8w_quantization_config(
+            is_per_channel=per_channel_quantization
+        ),
+        tosa_extensions=["int16"],
```
👍

Ninja91 commented on backends/arm/test/ops/test_linear.py:
```diff
@@ -258,3 +261,34 @@ def test_linear_vgf_INT(test_data: torch.Tensor):
+
+@pytest.mark.xfail(
+    reason="TOSA backend has limited INT16 support - view operations only support INT8/INT32/FP32/BOOL"
+)
```
This test is not failing but since we have not validate with tosa reference model run, I kept the xfail. Removing it in latest update to see if CI fails.

digantdesai commented:
> Addressed comments @per & @digantdesai . Removed unnecessary flaky marker and requirement to add flaky in pytest.ini

@Ninja91 there are still a couple of unresolved comments. Otherwise looks good to me.

Ninja91 commented on backends/arm/quantizer/arm_quantizer.py:
```diff
@@ -143,6 +143,111 @@ def get_symmetric_quantization_config(
 
 
+@functools.lru_cache
+def get_16a8w_quantization_config(
```
Done.

digantdesai reviewed (approved):
Thanks! Left a couple of nit comments.

Ninja91 commented:
GH issue: https://github.com/pytorch/executorch/issues/13635

Ninja91 committed:
Add 16A8W quantization configuration utility for ARM backend (#13641)

Summary:

This diff implements a 16A8W (16-bit activations, 8-bit weights) quantization configuration utility for the ExecutorTorch ARM backend, following the feedback from D79746479.

## Key Changes

**1. New Quantization Configuration Function**
- Add `get_16a8w_quantization_config()` in `fbcode/executorch/backends/arm/quantizer/arm_quantizer.py`
- Provides 16-bit activations with HistogramObserver (better precision than 8A8W)
- Maintains 8-bit weights with MinMaxObserver/PerChannelMinMaxObserver (memory efficient)
- **Technically supported by TOSA through [EXT-INT16 extension/profile](https://www.mlplatform.org/tosa/tosa_spec.html#_conv2d)**

## Benefits
- **Better Precision**: 16-bit activations provide higher precision than 8-bit. Useful for carrying precision for recurring neural nets.
ghstack-source-id: 305459598
exported-using-ghexport

Reviewed By: 3l1

Differential Revision: D79763381

Ninja91 committed:
Add 16A8W linear ops support and test (#13448)

Summary:

- Adds linear ops test using the 16A8W config in INT16 profile.
- Adds support in view ops validation for INT16 Dtype.
- Validated with TOSA pipeline test.
- Checked earlier marked flaky tests no longer flaky and remove markers.

Note: Not verified with tosa reference model run.

Reviewed By: digantdesai

Differential Revision: D80308822

Ninja91 commented:
@digantdesai  @per strangely seeing https://github.com/pytorch/executorch/actions/runs/17219418500/job/48850992655 error which doesn't look related. Am i missing something here?

Ninja91 commented:
Updated in a new PR: https://github.com/pytorch/executorch/pull/13658. Closing this.
